node/src/components/consensus/protocols/highway/round_success_meter.rs

```diff
@@ -0,0 +1,146 @@
+        self.proposals = Vec::new();
+    }
+
+    fn check_proposals_success(&self, state: &State<C>, proposal_h: &C::Hash) -> bool {
+        let total_w = state.total_weight();
+
+        let finality_detector =
+            FinalityDetector::<C>::new(max(total_w / 100 * THRESHOLD, Weight(1)));
+
+        // check for the existence of a level-1 summit
+        finality_detector.find_summit(1, proposal_h, state) == 1
+    }
+
+    /// Registers a proposal within this round - if it's finalized within the round, the round will
+    /// be successful.
+    pub fn new_proposal(&mut self, proposal_h: C::Hash, timestamp: Timestamp) {
+        // only add proposals from within the current round
+        if timestamp.millis() >= self.current_round_id
+            && timestamp.millis() - self.current_round_id < self.current_round_length
+        {
+            self.proposals.push(proposal_h);
+        }
+    }
+
+    /// If the current timestamp indicates that the round has ended, checks the known proposals for
+    /// a level-1 summit.
+    /// If there is a summit, the round is considered successful. Otherwise, it is considered
+    /// failed.
+    /// Next, a number of last rounds are being checked for success and if not enough of them are
+    /// successful, we return a higher round exponent for the future.
+    /// If the exponent shouldn't grow, and the round ID is divisible by a certain number, a lower
+    /// round exponent is returned.
+    pub fn calculate_new_exponent(&mut self, state: &State<C>) -> u8 {
+        let now = Timestamp::now();
+        // if the round hasn't finished, just return whatever we have now
+        if now.millis().saturating_sub(self.current_round_id) < self.current_round_length {
+            return self.new_exponent();
+        }
+
+        let current_round_index = self.current_round_id / self.current_round_length;
+        let new_round_index = now.millis() / self.current_round_length;
+
+        if mem::replace(&mut self.proposals, Vec::new())
+            .into_iter()
+            .any(|proposal| self.check_proposals_success(state, &proposal))
+        {
+            self.rounds.push_front(true);
+        } else {
+            self.rounds.push_front(false);
+        }
+
+        // if we're just switching rounds and more than a single round has passed, all the
+        // rounds since the last registered round have failed
+        for _ in 0..(new_round_index - current_round_index - 1) {
+            self.rounds.push_front(false);
+        }
+
+        self.current_round_id = new_round_index * self.current_round_length;
+
+        self.clean_old_rounds();
+
+        let new_exp = self.new_exponent();
+
+        if new_exp != self.current_round_exp {
+            self.change_exponent(new_exp, now);
+        }
+
+        new_exp
+    }
+
+    fn clean_old_rounds(&mut self) {
+        while self.rounds.len() > NUM_ROUNDS_TO_CONSIDER {
+            self.rounds.pop_back();
+        }
```

Comment:
Since `self.rounds` has always capacity of `NUM_ROUNDS_TO_CONSIDER`, and that is a constant, is there no method on `VecDeque` that clears it entirely?

Reply:
We don't want to clear it entirely here, though. We only want to clear the oldest entries that are more than `NUM_ROUNDS_TO_CONSIDER` rounds old.

Reply:
If we always consider `NUM_ROUNDS_TO_CONSIDER` when deciding whether to change round exponent or not, why do we need to keep more elements that that?

Reply:
We don't, but we might temporarily when appending a new success/failure to the queue. If we already hold `NUM_ROUNDS_TO_CONSIDER` elements and handle a new success/failure, we would hold one more element for a while - but then we will call this method and truncate the queue again.

Reply:
So you're saying that it's always `NUM_ROUNDS_TO_CONSIDER` except for the time when we add 1 element to it only to remove elements `1..=41`?

Reply:
No, no. If we add element 41, we will only remove element 0 - so that we will keep 40 elements at all times (until we change the round exponent). If for some reason we add multiple elements, say 41, 42 and 43 (I'm not sure if it's still possible - it was in a previous version), then we would remove elements 0, 1 and 2, so that the total is still 40.
So basically we always keep up to 40 elements, except for the brief moment right after adding an element, which can bring it over 40 for a while before we remove the older elements.
(Also, actually we add new elements in the front, not in the back, so the indices will be slightly different, but you get the idea.)

Reply:
Right, thanks :facepalm: